pkg/apiserver/profile/model.go

```diff
@@ -0,0 +1,121 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package profile
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/pingcap-incubator/tidb-dashboard/pkg/dbstore"
+)
+
+// TaskState is used to represent the task/task group state.
+type TaskState int
+
+// Built-in task state
+const (
+	TaskStateError TaskState = iota
+
+	// TaskGroup can only have these two states.
+	TaskStateRunning
+	TaskStateFinish
+)
+
+type TaskModel struct {
+	ID          uint      `json:"id" gorm:"primary_key"`
+	TaskGroupID uint      `json:"task_group_id" gorm:"index"`
+	State       TaskState `json:"state" gorm:"index"`
+	Addr        string    `json:"address"`
+	FilePath    string    `json:"file_path" gorm:"type:text"`
+	Component   string    `json:"component"`
+	CreateTime  int64     `json:"create_time"`
+	FinishTime  int64     `json:"finish_time"`
+	Error       string    `json:"error" gorm:"type:text"`
+}
+
+func (TaskModel) TableName() string {
+	return "profiling_tasks"
+}
+
+type TaskGroupModel struct {
+	ID    uint      `json:"id" gorm:"primary_key"`
+	State TaskState `json:"state" gorm:"index"`
+}
+
+func (TaskGroupModel) TableName() string {
+	return "profiling_task_groups"
+}
+
+func autoMigrate(db *dbstore.DB) error {
+	return db.AutoMigrate(&TaskModel{}).
+		AutoMigrate(&TaskGroupModel{}).
+		Error
+}
+
+// Task is the unit to fetch profiling information.
+type Task struct {
+	*TaskModel
+	db     *dbstore.DB
+	cancel context.CancelFunc
+}
+
+// NewTask creates a new profiling task.
+func NewTask(db *dbstore.DB, id uint, component, addr string) *Task {
+	return &Task{
+		TaskModel: &TaskModel{
+			TaskGroupID: id,
+			State:       TaskStateRunning,
+			Addr:        addr,
+			Component:   component,
+			CreateTime:  time.Now().Unix(),
+		},
+		db: db,
+	}
+}
+
+func (t *Task) run() {
+	ctx, cancel := context.WithCancel(context.Background())
+	t.cancel = cancel
```

Comment:
`ctx` and `cancel` need to be created and stored before run, otherwise `t.cancel` and `t.stop()` may have race condition (`t` is created, inserted into sync map, called `stop()` by other one before it `run()`)